### PR TITLE
add ignore closures option

### PR DESCRIPTION
**Pull request by kevinkreiser**
this adds a new costing option to ignore closures. recently we had some prs fixing the issue of allowing starting on a closure.

this pr does the following:

1. allows you to say you want to ignore closures
2. refactors the filter methods to only require the edge and the tile rather than the ids and the tiles (this reverts some of the changes from: https://github.com/valhalla/valhalla/pull/2597)
3. updates GetSpeed method on GraphTile to ignore closed speeds because they make costing go crazy

### CODE REVIEW

File: src/loki/reach.cc [hunk 1]
```diff
@@ -49,22 +49,15 @@ directed_reach Reach::operator()(const DirectedEdge* edge,
...
     if (!reader.GetGraphTile(node_id, tile))
       continue;
-    const auto* node = tile->node(node_id);
-
-    GraphId edge_id = tile->id();
-    edge_id.set_id(node->edge_index());
-
-    auto directed_edges = tile->GetDirectedEdges(node_id);
-    for (const auto* edge = directed_edges.begin(); edge != directed_edges.end(); ++edge, ++edge_id) {
+    for (const auto& edge : tile->GetDirectedEdges(node_id)) {
```
Comment (kevinkreiser) [inline src/loki/reach.cc:52]: revert from #2597

File: src/loki/reach.cc [hunk 2]
```diff
@@ -49,22 +49,15 @@ directed_reach Reach::operator()(const DirectedEdge* edge,
...
       // TODO: we'd rather say !edge.end_simple_restriction() and not !edge.restrictions()
       // TODO: but we'd need the predecessor information to do that so we punt 1 edge earlier
       // NOTE: we can go through the start of the restriction because only the end would mark a
       // potential stopping point (maybe a path followed the restriction)
 
       // if this edge is traversable we enqueue its end node
-      if (costing->Filter(edge, edge_id, tile) > 0 && !edge->end_restriction() &&
-          !edge->restrictions())
-        enqueue(edge->endnode(), reader, costing, tile);
+      if (costing->Filter(&edge, tile) > 0 && !edge.end_restriction() && !edge.restrictions())
```
Comment (kevinkreiser) [inline src/loki/reach.cc:59]: revert from #2597

File: src/loki/reach.cc [hunk 3]
```diff
@@ -96,18 +89,15 @@ directed_reach Reach::operator()(const DirectedEdge* edge,
       // get the opposing edge
       if (!reader.GetGraphTile(edge.endnode(), tile))
         continue;
+      const auto* node = tile->node(edge.endnode());
+      const auto* opp_edge = tile->directededge(node->edge_index() + edge.opp_index());
```
Comment (kevinkreiser) [inline src/loki/reach.cc:93]: revert from #2597

File: src/loki/search.cc [hunk 4]
```diff
@@ -475,10 +475,8 @@ struct bin_handler_t {
     // filtered then the reaches of both edges are the same
 
     const DirectedEdge* opp_edge = nullptr;
-    GraphId opp_edgeid;
-    if (reach.outbound > 0 && reach.inbound > 0 &&
-        (opp_edgeid = reader.GetOpposingEdgeId(edge_id, opp_edge, tile)) &&
-        costing->Filter(opp_edge, opp_edgeid, tile) > 0.f) {
+    if (reach.outbound > 0 && reach.inbound > 0 && (opp_edge = reader.GetOpposingEdge(edge, tile)) &&
```
Comment (kevinkreiser) [inline src/loki/search.cc:478]: revert from #2597

File: src/loki/search.cc [hunk 5]
```diff
@@ -779,7 +777,7 @@ struct bin_handler_t {
     // give back all the results
     return searched;
   }
-};
+}; // namespace
```
Comment (kevinkreiser) [inline src/loki/search.cc:780]: crap

File: valhalla/baldr/graphtile.h [hunk 6]
```diff
@@ -521,9 +524,11 @@ class GraphTile {
     if ((flow_mask & kCurrentFlowMask) && traffic_tile()) {
       auto directed_edge_index = std::distance(const_cast<const DirectedEdge*>(directededges_), de);
       auto volatile& live_speed = traffic_tile.trafficspeed(directed_edge_index);
-      if (live_speed.valid()) {
+      // only use current speed if its valid and non zero, a speed of 0 makes costing values crazy
+      uint8_t overall_speed;
+      if (live_speed.valid() && (overall_speed = live_speed.get_overall_speed()) > 0) {
```
Comment (kevinkreiser) [inline valhalla/baldr/graphtile.h:529]: prevent costing from getting a speed of 0

File: valhalla/sif/dynamiccost.h [hunk 7]
```diff
@@ -896,9 +895,8 @@ class DynamicCost {
    * @param  edgeid         GraphId of the opposing edge.
    * @return  Returns true if the edge is closed due to live traffic constraints, false if not.
    */
-  inline virtual bool IsClosedDueToTraffic(const baldr::GraphId& edgeid,
-                                           const baldr::GraphTile* tile) const {
-    return (flow_mask_ & baldr::kCurrentFlowMask) && tile->IsClosedDueToTraffic(edgeid);
+  inline virtual bool IsClosed(const baldr::DirectedEdge* edge, const baldr::GraphTile* tile) const {
+    return !ignore_closures_ && (flow_mask_ & baldr::kCurrentFlowMask) && tile->IsClosed(edge);
```
Comment (kevinkreiser) [inline valhalla/sif/dynamiccost.h:899]: its only closed if you care about closures

**Review by PureW — APPROVED** [review]

File: src/sif/bicyclecost.cc [hunk 8]
```diff
@@ -399,9 +398,7 @@ class BicycleCost : public DynamicCost {
    * mode used by the costing method. It's also used to filter
    * edges not usable / inaccessible by bicycle.
    */
-  float Filter(const baldr::DirectedEdge* edge,
-               const baldr::GraphId& /*edgeid*/,
-               const baldr::GraphTile* /*tile*/) const override {
+  float Filter(const baldr::DirectedEdge* edge, const baldr::GraphTile*) const override {
```
Comment (genadz) [inline src/sif/bicyclecost.cc:401]: that's great! I didn't notice that we `edgeid` is redundant here . it makes the code much easier)

**Review by genadz — APPROVED** [review]

File: src/sif/bicyclecost.cc (hunk 8 above)
Comment (kevinkreiser) [reply src/sif/bicyclecost.cc:401]: at first, i didnt think of it either or i would have suggested it. sorry for not coming up with it sooner actually!

**Pull request merged**